docs/design-documents/20231117-better-processors.md

```diff
@@ -0,0 +1,889 @@
+# Better Processors
+
+## Introduction
+
+### Background
+
+Conduit processors are currently pretty basic. We would like to make them more powerful
+to give users more freedom when manipulating in-flight data.
+
+Initially, processors were inspired by single message transforms (SMTs) provided by
+Kafka Connect. We tried to mimic the behavior as well as processor names to bring them
+closer to new Conduit users who had prior experience with Kafka Connect. A couple of
+issues arose because of this. The main issue is that Conduit is working with a single
+OpenCDC record that combines the key and payload, whereas Kafka Connect has separate
+payloads for both. The OpenCDC record additionally provides fields like metadata, position,
+operation, and two fields for the payload (before and after). Conduit also differentiates
+between raw data and structured data. All of this meant we tried to jam a square peg into
+a round hole and ended up with sub-par processors.
+
+This document tries to address these issues holistically and create a better experience
+around processors.
+
+### Goals
+
+- Keep existing processors to preserve backward compatibility.
+  - Old processors should only be deprecated but continue to work.
+- Add the ability to run a processor only on records that match a certain condition.
```

Comment:
i.e. we will have a "quick" way to filter out records (through the method proposed, a template) and then it's also possible to write a standalone processor which can filter out records as well?